src/client/transport.rs

```diff
@@ -1,241 +1,208 @@
-                            self.last_pong_received = Instant::now();
-                        }
-                    }
-
-                    _ => (),
-                }
-
-                Ok(Async::Ready(Some(msg)))
-            }
+        if let Poll::Ready(_) = Pin::new(&mut self.as_mut().ping_interval).poll_next(cx) {
+            self.as_mut().send_ping()?;
+            self.as_mut().set_deadline();
         }
+
+        Poll::Pending
     }
 }
 
-impl<T> Sink for IrcTransport<T>
+/// An IRC transport that handles core functionality for the IRC protocol. This is used in the
+/// implementation of `Connection` and ultimately `IrcServer`, and plays an important role in
+/// handling connection timeouts, message throttling, and ping response.
+pub struct Transport<T> {
+    /// The inner connection framed with an `IrcCodec`.
+    inner: Framed<T, IrcCodec>,
+    /// Helper for handle pinging.
+    pinger: Option<Pinger>,
+}
+
+impl<T> Unpin for Transport<T> where T: Unpin {}
+
+impl<T> Transport<T>
 where
-    T: AsyncRead + AsyncWrite,
+    T: Unpin + AsyncRead + AsyncWrite,
 {
-    type SinkItem = Message;
-    type SinkError = error::IrcError;
-
-    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
-        // If the ping timeout has been reached, we close the connection out and return an error.
-        if self.ping_timed_out() {
-            self.close()?;
-            return Err(error::IrcError::PingTimeout)
-        }
+    /// Creates a new `Transport` from the given IRC stream.
+    pub fn new(
+        config: &Config,
+        inner: Framed<T, IrcCodec>,
+        tx: UnboundedSender<Message>,
+    ) -> Transport<T> {
+        let pinger = Some(Pinger::new(tx, config));
+
+        Transport { inner, pinger }
+    }
 
-        // Check if the oldest message in the rolling window is discounted.
-        if let Async::Ready(()) = self.rolling_burst_window_front()? {
-            self.current_burst_messages -= 1;
-            self.rolling_burst_window.pop_front();
-        }
+    /// Gets the inner stream underlying the `Transport`.
+    pub fn into_inner(self) -> Framed<T, IrcCodec> {
+        self.inner
+    }
+}
 
-        // Throttling if too many messages have been sent recently.
-        if self.current_burst_messages >= self.max_burst_messages {
-            // When throttled, we know we need to finish sending what's already queued up.
-            self.poll_complete()?;
-            return Ok(AsyncSink::NotReady(item))
-        }
+impl<T> Stream for Transport<T>
+where
+    T: Unpin + AsyncRead + AsyncWrite,
+{
+    type Item = Result<Message, error::Error>;
 
-        match self.inner.start_send(item)? {
-            AsyncSink::NotReady(item) => Ok(AsyncSink::NotReady(item)),
-            AsyncSink::Ready => {
-                self.current_burst_messages += 1;
-                self.rolling_burst_window.push_back(self.burst_timer.sleep(Duration::from_secs(
```

Comment:
Why did you completely wipe out all the flood control settings?

Reply:
Ah. Yeah, it hasn't been ported yet (see TODO above).

I also wanted to pitch not including it since plugging in an external rate limiter is fairly straight forward. For example, I've written and use https://docs.rs/leaky-bucket/0.4.1/leaky_bucket/

Alternatively if this can be put into a higher level abstraction so me as a user can decide whether or not to include it, if you want to have access to just the protocol bits.

Either way, I'm not sure I have time to port it right now. But you can always delay a release until it has been added again.